datasource/src/test/scala/examples/BufferTiles.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * This software is licensed under the Apache 2 license, quoted below.
+ *
+ * Copyright 2020 Astraea, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *     [http://www.apache.org/licenses/LICENSE-2.0]
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package examples
+
+import geotrellis.raster.mapalgebra.focal.Square
+import org.apache.spark.sql._
+import org.locationtech.rasterframes._
+import org.locationtech.rasterframes.datasource.raster._
+import org.locationtech.rasterframes.tiles.ProjectedRasterTile
+
+object BufferTiles extends App {
+
+  implicit val spark =
+    SparkSession
+      .builder()
+      .master("local[*]")
+      .appName("RasterFrames")
+      .withKryoSerialization
+      .getOrCreate()
+      .withRasterFrames
+
+  spark.sparkContext.setLogLevel("ERROR")
+
+  import spark.implicits._
+
+  val example = "https://raw.githubusercontent.com/locationtech/rasterframes/develop/core/src/test/resources/LC08_B7_Memphis_COG.tiff"
+
+  val tile =
+    spark
+      .read
+      .raster
+      .from(example)
+      .withBufferSize(1)
+      .withTileDimensions(100, 100)
+      .load()
+      .limit(1)
+      .select($"proj_raster")
+      .select(rf_focal_max($"proj_raster", Square(1)))
+      // .select(rf_aspect($"proj_raster"))
+      // .select(rf_hillshade($"proj_raster", 315, 45, 1))
+      .as[Option[ProjectedRasterTile]]
+      // .show(false)
+      .first()
+
+  // tile.get.renderPng().write("/tmp/hillshade-buffered.png")
+  // tile.get.renderPng().write("/tmp/hillshade-nobuffered.png")
+
+  // spark.stop()
+}
```

Comment:
Wow....  this is fantastic work!!!

Reply:
How did you do this without substantially changing `TileUDT`? Where is the buffer size encoded in the dataframe?

Reply:
@metasim the it is a part of the RasterRef now, only raster refs for now can contain bufferSize; RasterRef is a case class.

Reply:
@metasim check how we can flatten inputs and extract the BufferTile https://github.com/locationtech/rasterframes/pull/563/files#diff-f3a16be8b11d3ca1fd027d55f7ab7e9f06e4ccb312ea0e4760d1361016ae9fa5R57-R66 Later it is possible to pattern match on it. 

RasterRef operates by default with BufferTiles, the default buffer size = 0.

Reply:
To improve the seamless integraion we can add the buffer size into the TileUDT tbh, that can extend all ops to work with the explicitly BufferTile object passed into the DataFrame.

Reply:
> extend all ops to work with the explicitly BufferTile object passed into the DataFrame

Can we do this without every Expression having to have this logic?

Reply:
> @metasim check how we can flatten inputs and extract the BufferTile https://github.com/locationtech/rasterframes/pull/563/files#diff-f3a16be8b11d3ca1fd027d55f7ab7e9f06e4ccb312ea0e4760d1361016ae9fa5R57-R66 Later it is possible to pattern match on it.

Not sure this link is working still..


> RasterRef operates by default with BufferTiles, the default buffer size = 0.

And I'm wondering: what if all of RasterFrames worked this way?